registry/distribution/repository.go

```diff
@@ -0,0 +1,674 @@
+// manifest descriptor.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md
+func (r *Repository) UpEdges(ctx context.Context, desc ocispec.Descriptor) ([]ocispec.Descriptor, error) {
+	var res []ocispec.Descriptor
+	if err := r.Referrers(ctx, desc, func(referrers []artifactspec.Descriptor) error {
+		for _, referrer := range referrers {
+			res = append(res, descriptor.ArtifactToOCI(referrer))
+		}
+		return nil
+	}); err != nil {
+		return nil, err
+	}
+	return res, nil
+}
+
+// Referrers returns the manifest descriptors directly referencing the given
+// manifest descriptor.
+// Reference: https://github.com/oras-project/artifacts-spec/blob/main/manifest-referrers-api.md
+func (r *Repository) Referrers(ctx context.Context, desc ocispec.Descriptor, fn func(referrers []artifactspec.Descriptor) error) error {
+	// TODO(shizhMSFT): filter artifact type
+	url := fmt.Sprintf("%s/manifests/%s/referrers", r.artifactEndpoint(), desc.Digest)
+	if r.ReferrerListPageSize > 0 {
+		url = fmt.Sprintf("%s?n=%d", url, r.ReferrerListPageSize)
+	}
+
+	var err error
+	for err == nil {
+		url, err = r.referrers(ctx, desc, fn, url)
+	}
+	if err != errNoLink {
+		return err
+	}
+	return nil
+}
+
+// referrers returns a single page of the manifest descriptors directly
+// referencing the given manifest descriptor with the next link.
+func (r *Repository) referrers(ctx context.Context, desc ocispec.Descriptor, fn func(referrers []artifactspec.Descriptor) error, url string) (string, error) {
+	req, err := http.NewRequestWithContext(ctx, http.MethodGet, url, nil)
+	if err != nil {
+		return "", err
+	}
+
+	resp, err := r.Client.Do(req)
+	if err != nil {
+		return "", err
+	}
+	defer resp.Body.Close()
+
+	if resp.StatusCode != http.StatusOK {
+		return "", parseErrorResponse(resp)
+	}
+
+	var page struct {
+		References []artifactspec.Descriptor `json:"references"`
+	}
+	lr := limitReader(resp.Body, r.MaxMetadataBytes)
+	if err := json.NewDecoder(lr).Decode(&page); err != nil {
+		return "", err
+	}
+	if err := fn(page.References); err != nil {
+		return "", err
+	}
+
+	return parseLink(resp)
+}
+
+// scheme returns HTTP scheme used to access the remote registry.
+func (r *Repository) scheme() string {
+	if r.PlainHTTP {
+		return "http"
+	}
+	return "https"
+}
+
+// endpoint returns the base endpoint of the remote registry.
+func (r *Repository) endpoint() string {
+	return fmt.Sprintf("%s://%s/v2/%s", r.scheme(), r.Reference.Host(), r.Reference.Repository)
+}
+
+// artifactEndpoint returns the base endpoint of the remote registry for ORAS
+// artifacts.
+func (r *Repository) artifactEndpoint() string {
+	return fmt.Sprintf("%s://%s/oras/artifacts/v1/%s", r.scheme(), r.Reference.Host(), r.Reference.Repository)
+}
+
+// delete removes the content identified by the descriptor in the entity "blobs"
+// or "manifests".
+func (r *Repository) delete(ctx context.Context, entity string, target ocispec.Descriptor) error {
```

Comment:
Should we make this an  `iota` since only 2 values are acceptable.

Reply:
Let's keep it simple since the `delete` is a private method.

Reply:
But would also image it makes for lesser CPU cycles in general wouldn’t it?

Reply:
Well, that's `true`.

Although `string` values are passed by values, they are actually passing by reference since `string` is a [StringHeader](https://pkg.go.dev/reflect#StringHeader) struct and string copy is shadow copy.

```golang
type StringHeader struct {
	Data uintptr
	Len  int
}
```

Definitely, an enum will be faster than string. Since it has only 2 accepatbale values, let's make it a `bool`.